src/pip/_internal/req/req_install.py

```diff
@@ -345,7 +346,7 @@ def ensure_build_location(
 
         # When parallel builds are enabled, add a UUID to the build directory
         # name so multiple builds do not interfere with each other.
-        dir_name: str = canonicalize_name(self.name)
+        dir_name: str = canonicalize_name(self.req.name)
```

Comment:
Is this correct? IIRC these two have a slight sematic difference when a non-standard requirement is passed in (unnamed, directory path etc.)

Reply:
Yes, it's correct. `name` is a property: https://github.com/pypa/pip/blob/339a31b5fc3957710372091fdee1f0c63148a9f6/src/pip/_internal/req/req_install.py#L225
We know `self.req` is not None from the return above (and now that we remove no-strict-optional, mypy would complain if it could be None)